File: booking_service/src/core/usecase/delete_unit_usecase.go
package usecase

import (
	"booking_service/core/port"
	"booking_service/kernel/utils"
	"context"
	"github.com/golibs-starter/golib/log"
)

type IDeleteUnitUseCase interface {
	DeleteUnit(ctx context.Context, accID, unitID int64) error
}

type DeleteUnitUseCase struct {
	unitPort             port.IUnitPort
	cachePort            port.ICachePort
	dbTransactionUseCase IDatabaseTransactionUseCase
}

func (d DeleteUnitUseCase) DeleteUnit(ctx context.Context, accID, unitID int64) error {
	_, err := d.unitPort.GetUnitByAccIDAndID(ctx, accID, unitID)
	if err != nil {
		log.Error(ctx, "get unit by id failed ", err)
		return err
	}

	tx := d.dbTransactionUseCase.StartTransaction()
	defer func() {
		if errRollBack := d.dbTransactionUseCase.Rollback(tx); errRollBack != nil {
			log.Error(ctx, "Rollback delete unit failed, : ", errRollBack)
		} else {
			log.Info(ctx, "Rollback delete unit success")
		}
	}()

	err = d.unitPort.DeleteUnitByID(ctx, tx, unitID)
	if err != nil {
		log.Error(ctx, "delete unit failed, : ", err)
		return err
	}

	errCommit := d.dbTransactionUseCase.Commit(tx)
	if errCommit != nil {
		log.Error(ctx, "Commit delete unit failed, : ", errCommit)
		return errCommit
	}

	// clear cache
	go func() {
		err := d.cachePort.DeleteFromCache(ctx, utils.BuildCacheKeyGetAccommodation(accID))
		if err != nil {
			log.Error(ctx, "delete unit from cache failed", err)
		}
	}()

	return nil
}

func NewDeleteUnitUseCase(unitPort port.IUnitPort, dbTransactionUseCase IDatabaseTransactionUseCase,
	cachePort port.ICachePort) IDeleteUnitUseCase {
	return &DeleteUnitUseCase{
		unitPort:             unitPort,
		cachePort:            cachePort,
		dbTransactionUseCase: dbTransactionUseCase,
	}
}

File: accommodation_service/src/main/java/huy/project/accommodation_service/core/usecase/DeleteUnitUseCase.java
package huy.project.accommodation_service.core.usecase;

import huy.project.accommodation_service.core.domain.constant.ErrorCode;
import huy.project.accommodation_service.core.domain.constant.TopicConstant;
import huy.project.accommodation_service.core.domain.kafka.UnitDeletedEvent;
import huy.project.accommodation_service.core.exception.AppException;
import huy.project.accommodation_service.core.port.ICachePort;
import huy.project.accommodation_service.core.port.IKafkaPublisher;
import huy.project.accommodation_service.core.port.IUnitPort;
import huy.project.accommodation_service.kernel.utils.CacheUtils;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Slf4j
public class DeleteUnitUseCase {
    private final IUnitPort unitPort;
    private final ICachePort cachePort;
    private final IKafkaPublisher kafkaPublisher;

    private final GetAccommodationUseCase getAccommodationUseCase;

    @Transactional(rollbackFor = Exception.class)
    public void deleteUnit(Long userId, Long accId, Long unitId) {
        validateRequest(userId, accId, unitId);

        var unit = unitPort.getUnitByAccIdAndId(accId, unitId);
        unit.setIsDeleted(true);

        unitPort.save(unit);

        // clear cache
        cachePort.deleteFromCache(CacheUtils.buildCacheKeyGetAccommodationById(accId));

        // publish event to kafka
        publishEventToKafka(accId, unitId);
    }

    private void validateRequest(Long userId, Long accId, Long unitId) {
        var accommodation = getAccommodationUseCase.getAccommodationById(accId);
        if (accommodation == null) {
            log.error("Accommodation with id {} not found", accId);
            throw new AppException(ErrorCode.ACCOMMODATION_NOT_FOUND);
        }
        if (!accommodation.getHostId().equals(userId)) {
            log.error("Accommodation with id {} not belong to user {}", accId, userId);
            throw new AppException(ErrorCode.FORBIDDEN_DELETE_UNIT);
        }

        var unit = unitPort.getUnitByAccIdAndId(accId, unitId);
        if (unit == null) {
            log.error("Unit with id {} not found in accommodation {}", unitId, accId);
            throw new AppException(ErrorCode.UNIT_NOT_FOUND);
        }
        if (unit.getIsDeleted() != null && unit.getIsDeleted()) {
            log.error("Unit with id {} already deleted", unitId);
            throw new AppException(ErrorCode.UNIT_ALREADY_DELETED);
        }
    }

    private void publishEventToKafka(Long accId, Long unitId) {
        var event = UnitDeletedEvent.builder()
                .accommodationId(accId)
                .unitId(unitId)
                .build();
        kafkaPublisher.pushAsync(event.toDomainEventDto(), TopicConstant.AccommodationEvent.TOPIC, null);
    }
}
